LOOP-1704 Allow glucose data without HealthKit write access

darinkrauss opened the pull request:
- https://tidepool.atlassian.net/browse/LOOP-1704
- Refactor GlucoseStore to make Core Data source of truth and HealthKit optional
- Refactor various glucose-related types to support GlucoseStore changes
- Minor renaming for consistency between stores
- Add tests

Commit:
LOOP-1704 Allow glucose data without HealthKit write access

- https://tidepool.atlassian.net/browse/LOOP-1704
- Refactor GlucoseStore to make Core Data source of truth and HealthKit optional
- Refactor various glucose-related types to support GlucoseStore changes
- Minor renaming for consistency between stores
- Add tests

rickpasetto commented on LoopKit/GlucoseKit/GlucoseStore.swift:
```diff
@@ -143,511 +150,496 @@ public final class GlucoseStore: HealthKitSampleStore {
+    ///   - completion: A closure called once the glucose samples have been retrieved.
+    ///   - result: An array of glucose samples, in chronological order by startDate, or error.
+    public func getGlucoseSamples(start: Date? = nil, end: Date? = nil, completion: @escaping (_ result: Result<[StoredGlucoseSample], Error>) -> Void) {
+        queue.async {
```
[nit] This is a small nit observation, and I may not have all the pieces understood here, but it looks like the private `getGlucoseSamples` below blocks in `performAndWait`, and then results are passed through `completion` here.  Seems fine and should work, but won't it potentially block `queue`?  Why not just pass the `completion` to it, and have the `getGlucoseSamples` function below do a regular `perform` and not hop queues at all?  

Again, I may not have the whole picture here, and it is a minor comment.

darinkrauss replied:
The reason for the queues in each of these stores is to synchronize access to associated data (transient properties on the store, but not in Core Data) **with and without** changes to Core Data. We don't want to necessarily use the managed object context queue for that purpose since the managed object context is shared between a number of stores.

rickpasetto reviewed (approved):
LGTM.  One minor comment, but other than that all good!

ps2 commented on LoopKit/GlucoseKit/GlucoseStore.swift:
```diff
@@ -51,7 +46,7 @@ public enum GlucoseStoreResult<T> {
 public final class GlucoseStore: HealthKitSampleStore {
 
-    /// Notification posted when glucose samples were changed, either via add/replace/delete methods or from HealthKit
+    /// Notification posted when glucose samples were changed, either via direct add or from HealthKit
```
[nit] Is this signal changing to only include adds?  Or are replace/delete's still firing this signal?

darinkrauss replied:
Technically this is sent for *any* change to the GlucoseStore, but changes to the GlucoseStore can only happen on direct add or HealthKit update. There is no way to direct replace or delete glucose samples (only via HealthKit from an external app). The "deletion" that does occur is through purge (either when sample fall off the end of the cache or when an external app adds entries that this cache is also managing) and purging only occurs at the end of a direct add or HealthKit update. So, technically, glucose samples are only changed via direct add or HealthKit update (which may also purge, but that is an internal implementation detail.)

ps2 replied:
I'm just thinking about this from a caller's perspective.  If someone calls `deleteCarbEntry` or `replaceCarbEntry`, will this signal be fired? If so, I think we should document this notification as responding to add, replace, deletes, and adds or deletes from HK.

darinkrauss replied:
This is the glucose store, not the carb store. A user of the glucose store cannot explicitly replace or delete a glucose entry (and deletes only happen from a purge related to an add or HealthKit operation).

The carb store docs do say "either via add/replace/delete methods or from HealthKit" since a user of the carb store can explicitly replace or delete carbs.

ps2 replied:
Doh! In the wrong pew. :) Sorry, and thanks for pointing out my mistake.

ps2 commented on LoopKit/GlucoseKit/GlucoseStore.swift:
```diff
@@ -143,511 +150,496 @@ public final class GlucoseStore: HealthKitSampleStore {
+
+        let count = try cacheStore.managedObjectContext.count(for: request)
+        guard count == 0 else {
+            return false
```
[nit] Log this unexpected state? Or is it expected at times.

darinkrauss replied:
Added log in calling code.

ps2 commented on LoopKit/GlucoseKit/SyncGlucoseObject.swift:
```diff
@@ -0,0 +1,58 @@
+import Foundation
+import HealthKit
+
+public struct SyncGlucoseObject: Codable, Equatable {
```
Why do we need a new type for watch?  Seems like watch should use the same types that Loop uses. Looks very similar to StoredGlucoseSample

darinkrauss replied:
Yes, it is very similar to StoredGlucoseSample, but I chose to use a separate type for 2 reasons:
1. StoredGlucoseSample would require a manually implemented Codable since it uses an `HKQuantity` (which cannot directly be Codable).
2. For symmetry with CarbStore.
If you want me to use StoredGlucoseSample I can implement the custom Codable implementation. Let me know.

ps2 replied:
RE no. 1: I don't think this is a good reason to create a new type. We interact with the types all the time in code, and have to reason about them all the time; the serialization is more of a one time thing.

RE no. 2: I'm not sure I follow what symmetry you're referring to.

darinkrauss replied:
Pending results of other thread regarding WatchHistoricalGlucose I will remove SyncGlucoseObject and use custom Codable implementation of StoredGlucoseSample.

darinkrauss replied:
Removed SyncGlucoseObject and replaced with StoredGlucoseSample.

ps2 reviewed (commented):
Generally looks good to me. I have a question about SyncGlucoseObject and its purpose. Seems like it should not be needed, but I may be missing something.

Commit:
Updates based upon PR review

- Additional logging in CarbStore and GlucoseStore
- Fix update notification for CarbStore and GlucoseStore
- Minor renaming for clarity

Commit:
Updates from PR review

ps2 reviewed (approved):
LGTM! Just a doc comment; not needing re-review, and will be fine even if the doc suggestion isn't taken.

Commit:
Merge branch 'dev' into darinkrauss/LOOP-1704/glucose-core-data
